In [1]:
import numpy as np

1. Можно ли отобрать наиболее значимые признаки с помощью PCA?

Составленная матрица весов может быть показателем наиболее значимых признаков.

3*. Написать свою реализацию метода главных компонент с помощью сингулярного разложения с использованием функции numpy.linalg.svd()

In [26]:
def pca_svd(data):
    u, d , v = np.linalg.svd(data)

    pairs = list(zip(d, v))
    pairs.sort(key=lambda x: x[0], reverse=True)
    s_value_sum = sum(d)
    var_exp = [i/s_value_sum*100 for i in sorted(d, reverse=True)]
    print(np.cumsum(var_exp))
    l = input()
    l = int(l)
    W = []
    for i in range(l):
        W.append(pairs[i][1])
    W = np.array(W)
    return np.dot(data, W.T)

In [55]:
a = np.array([[1,2,3],[4,5,6],[7,8,9],[10,11,12]])

In [56]:
pca_svd(a)

[ 95.17565005 100.         100.        ]
2


array([[ -3.58705934,   1.06442721],
       [ -8.75770068,   0.55016253],
       [-13.92834202,   0.03589786],
       [-19.09898335,  -0.47836682]])

4*. Обучить любую модель классификации на датасете iris до применения PCA и после него. Сравнить качество классификации по отложенной выборке.

In [48]:
from sklearn import datasets

In [51]:
iris = datasets.load_iris()

In [52]:
X = iris.data
y = iris.target

In [57]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score
from sklearn.model_selection import train_test_split

In [60]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

In [61]:
model1 = RandomForestClassifier()

In [62]:
model1.fit(X_train, y_train)

D:\Anaconda3\lib\site-packages\sklearn\ensemble\forest.py:245: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)


RandomForestClassifier(bootstrap=True, class_weight=None, criterion='gini',
                       max_depth=None, max_features='auto', max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, n_estimators=10,
                       n_jobs=None, oob_score=False, random_state=None,
                       verbose=0, warm_start=False)

In [63]:
y_pred = model1.predict(X_test)

In [64]:
accuracy_score(y_test, y_pred)

1.0

In [65]:
X_pca = pca_svd(X)

[ 80.59340691  95.51028371  98.41699969 100.        ]
2


In [68]:
X_train_pca, X_test_pca, y_train_pca, y_test_pca = train_test_split(X_pca, y, test_size=0.3, random_state=42)

In [69]:
model2 = RandomForestClassifier()

In [70]:
model2.fit(X_train_pca, y_train_pca)

D:\Anaconda3\lib\site-packages\sklearn\ensemble\forest.py:245: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)


RandomForestClassifier(bootstrap=True, class_weight=None, criterion='gini',
                       max_depth=None, max_features='auto', max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, n_estimators=10,
                       n_jobs=None, oob_score=False, random_state=None,
                       verbose=0, warm_start=False)

In [71]:
y_pred_pca = model2.predict(X_test_pca)

In [72]:
accuracy_score(y_test_pca, y_pred_pca)

1.0

Поскольку метрика качества не меняется, делаем вывод что модель отлично предсказывает на новых данных из 2 признаков, объясняющих 95.5% дисперсии исходных данных.